src/dag/write.ts

```diff
@@ -0,0 +1,158 @@
+import {getRootAsMeta} from './meta.js';
+import type * as kv from '../kv/mod.js';
+import {HeadChange, toLittleEndian, fromLittleEndian} from './dag.js';
+import {chunkDataKey, chunkMetaKey, headKey, chunkRefCountKey} from './key.js';
+import {Read} from './read.js';
+import type {Chunk} from './chunk.js';
+import * as utf8 from '../utf8.js';
+
+export class Write {
+  private readonly _kvw: kv.Write;
+  private readonly _mutatedChunks = new Set<string>();
+  private readonly _changedHeads = new Map<string, HeadChange>();
+
+  constructor(kvw: kv.Write) {
+    this._kvw = kvw;
+  }
+
+  read(): Read {
+    return new Read(this._kvw);
+  }
+
+  async putChunk(c: Chunk): Promise<void> {
+    const {hash, data, meta} = c;
+    const key = chunkDataKey(hash);
+    const p1 = this._kvw.put(key, data);
+    let p2;
+    if (meta !== undefined) {
+      p2 = this._kvw.put(chunkMetaKey(hash), meta);
```

Comment:
Similarly let's ensure we always have a counts key present. If the semantics allow for putting an existing chunk (with same data and meta) that means we must check here whether a count exists and if not default it to zero.